### PR TITLE
Structs: Fields check

### PR DESCRIPTION
**Pull request by gzanitti**
Note: Like #1910, this PR also includes the same modifications to expression processing, in order to run tests.

### CODE REVIEW

File: pil-analyzer/tests/types.rs [hunk 1]
```diff
@@ -734,3 +734,39 @@ fn prover_functions() {
...
+    let input = "
+    struct Dot { x: int, y: int }
+    let f: int -> Dot = |i| Dot{a: 0, y: i};
+    let x = f(0);
+    ";
+    type_check(input, &[]);
+}
+
+#[test]
+#[should_panic = "Field 'z' is declared but never used in struct 'Point' initialization"]
```
Comment (chriseth) [inline pil-analyzer/tests/types.rs:761]: I would say something like "forgot field z" - please check what rustc's errors look like.

File: pil-analyzer/src/pil_analyzer.rs [hunk 2]
```diff
@@ -257,6 +258,69 @@ impl PILAnalyzer {
         }
     }
 
+    pub fn structs_check(&self) -> Result<(), Vec<Error>> {
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:261]: Every public function needs documentation.

File: pil-analyzer/src/pil_analyzer.rs [hunk 3]
```diff
@@ -56,6 +56,7 @@ fn analyze<T: FieldElement>(files: Vec<PILFile>) -> Result<Analyzed<T>, Vec<Erro
     let mut analyzer = PILAnalyzer::new();
     analyzer.process(files)?;
     analyzer.side_effect_check()?;
+    analyzer.structs_check()?;
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:59]: Maybe `struct_fields_check` to be more precise?

File: pil-analyzer/src/pil_analyzer.rs (hunk 2 above)
Comment (chriseth) [reply pil-analyzer/src/pil_analyzer.rs:261]: Ah wait, it doesn't need to be public. But please document it nevertheless :)

File: pil-analyzer/src/pil_analyzer.rs [hunk 4]
```diff
@@ -257,6 +258,69 @@ impl PILAnalyzer {
...
+                ))) = value
+                {
+                    Some((name.clone(), struct_decl.clone()))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        for (_, def) in self.definitions.values() {
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:278]: This is not enough, we also need to traverse proof items and impls. Maybe the easiest is to implement `Children<Expression>` on `PilAnalyzer` and then call `self.all_children()`.

File: pil-analyzer/tests/types.rs [hunk 5]
```diff
@@ -734,3 +734,39 @@ fn prover_functions() {
     ";
     type_check(input, &[("a", "", "int"), ("b", "", "()[]")]);
 }
+
```
Comment (chriseth) [inline pil-analyzer/tests/types.rs:737]: Please add a test like `enum A { X } let a = A{x: 8};`, i.e. where you use an enum for a struct expression.

File: pil-analyzer/src/pil_analyzer.rs [hunk 6]
```diff
@@ -257,6 +258,69 @@ impl PILAnalyzer {
...
+                    let Some(struct_decl) = struct_declarations.get(&struct_expr.name.to_string())
+                    else {
+                        errors.push(sr.with_error(format!(
+                            "Struct '{}' has not been declared",
+                            struct_expr.name
+                        )));
+                        return;
+                    };
+
+                    let mut used_fields = HashSet::new();
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:292]: What about something like
```
let declared_fields: HashSet<_> = struct_decl.fields.iter().map(|f| f.name).collect();
let used_fields: HashSet<_> = sturct_expr.fields...;
for f in declared_fields.difference(used_field) { ... }
for f in used_fields.difference(declared_fields) { ... }
```

Comment (chriseth) [reply pil-analyzer/src/pil_analyzer.rs:292]: The last two lines could even be
```
errors.extend(declared_fields.difference(used_field).map(|f| ... ))
```

**Comment by chriseth** [comment]
Oh and please also add a test where you mention a field twice.

File: pil-analyzer/tests/types.rs [hunk 7]
```diff
@@ -768,32 +768,68 @@ fn wrong_struct() {
...
+        y: int,
+    }
+    let x = A{ y: 0, x: 2, y: 1 };
+";
+
+    type_check(input, &[]);
+}
+
+#[test]
+#[should_panic = "Field 'x' is already declared"]
```
Comment (chriseth) [inline pil-analyzer/tests/types.rs:814]: Why is this a different error message than above?

File: pil-analyzer/src/pil_analyzer.rs [hunk 8]
```diff
@@ -257,6 +258,64 @@ impl PILAnalyzer {
         }
     }
 
+    /// Verifies that all struct instantiations match their corresponding declarations (existence of field names, completeness)
+    fn struct_fields_check(&self) -> Result<(), Vec<Error>> {
+        let mut errors = vec![];
+
+        self.all_children().for_each(|expr| {
+            if let Expression::StructExpression(sr, struct_expr) = expr {
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:266]: You could decrease the indentation here by either using `filter_map` or `for expr in ... { let Expression::StructExpression{ ... } else { continue };` or something else. This is good because when someone reads until here, it is good to know that there will be no `else` clause.

File: pil-analyzer/src/pil_analyzer.rs [hunk 9]
```diff
@@ -257,6 +258,64 @@ impl PILAnalyzer {
...
+                )) = self.definitions.get(&struct_expr.name.to_string())
+                else {
+                    panic!("Struct '{}' has not been declared", struct_expr.name);
+                };
+
+                let mut declared_fields = HashSet::new();
+                for field in &struct_decl.fields {
+                    if !declared_fields.insert(&field.name) {
+                        errors.push(
+                            sr.with_error(format!("Field '{}' is already declared", field.name)),
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:281]: Oh I see where this error message comes from. If you do it this way, then 1) we won't get an error for a struct declaration that is never used and 2) we will get the error at the location of the use and not the declaration and 3) we will get multiple errors for each use.

File: pil-analyzer/src/pil_analyzer.rs [hunk 10]
```diff
@@ -542,6 +601,24 @@ impl PILAnalyzer {
     }
 }
 
+impl AllChildren<Expression> for PILAnalyzer {
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:604]: If you just implement Children, will you get AllChildren for free?

Comment (chriseth) [reply pil-analyzer/src/pil_analyzer.rs:604]: If you implement `Children`, please take care to return the expression itself instead of calling children on them, because in the latter case, it will skip the expression itself.

File: pil-analyzer/tests/types.rs (hunk 7 above)
Comment (gzanitti) [reply pil-analyzer/tests/types.rs:814]: Both errors are for the case of repeated fields, but "specified more than once" is for the instantiation and "is already declared" is for the declaration. It's the way rust handles them.

File: pil-analyzer/src/pil_analyzer.rs [hunk 11]
```diff
@@ -257,6 +257,113 @@ impl PILAnalyzer {
...
+        let structs_exprs = self.all_children().filter_map(|expr| {
+            if let Expression::StructExpression(
+                sr,
+                StructExpression {
+                    name: Reference::Poly(poly),
+                    fields,
+                },
+            ) = expr
+            {
+                Some((sr, poly.name.clone(), fields))
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:274]: Please don't clone if not necessary.

File: pil-analyzer/src/pil_analyzer.rs [hunk 12]
```diff
@@ -257,6 +257,113 @@ impl PILAnalyzer {
...
+                let duplicate_fields: Vec<_> =
+                    fields.iter().map(|f| &f.name).duplicates().collect();
+
+                errors.extend(
+                    duplicate_fields
+                        .into_iter()
+                        .map(|f| sr.with_error(format!("Field '{f}' specified more than once"))),
+                );
+            } else {
+                errors.push(sr.with_error(format!("Struct '{name}' has not been declared")));
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:317]: Can this happen? Shouldn't it be a panic? Or should it rather be an error that we exected a struct and got something else?

File: pil-analyzer/src/pil_analyzer.rs [hunk 13]
```diff
@@ -257,6 +257,113 @@ impl PILAnalyzer {
...
+                ))
+            )
+        });
+
+        for (name, (symbol, def)) in structs_decls {
+            if !visited_structs.contains(name) {
+                errors.push(
+                    symbol
+                        .source
+                        .with_error(format!("warning: struct '{name}' is never constructed")), // TODO: Warningx
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:337]: I think it should be totally fine for a struct to never be constructed.

File: pil-analyzer/src/pil_analyzer.rs [hunk 14]
```diff
@@ -257,6 +257,113 @@ impl PILAnalyzer {
...
+                    .fields
+                    .iter()
+                    .map(|f| &f.name)
+                    .duplicates()
+                    .collect();
+
+                errors.extend(duplicate_declaration_fields.into_iter().map(|f| {
+                    symbol
+                        .source
+                        .with_error(format!("Field '{f}' is already declared",))
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:353]: Declared twice? Or do we have the source reference directly on the second field declaration?

Comment (gzanitti) [reply pil-analyzer/src/pil_analyzer.rs:353]: No, the source is for the complete struct. Maybe `is declared more than once` then?

File: ast/src/parsed/mod.rs [hunk 15]
```diff
@@ -42,6 +42,8 @@ pub enum SymbolCategory {
     TypeConstructor,
     /// A trait declaration, which can be used as a type.
     TraitDeclaration,
+    /// A struct
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:45]: If you introduce this and it cannot be unified with `Type` then please also rename `Type`.

File: ast/src/parsed/mod.rs [hunk 16]
```diff
@@ -54,6 +56,9 @@ impl SymbolCategory {
                 request == SymbolCategory::TypeConstructor || request == SymbolCategory::Value
             }
             SymbolCategory::TraitDeclaration => request == SymbolCategory::TraitDeclaration,
+            SymbolCategory::Struct => {
+                request == SymbolCategory::Struct || request == SymbolCategory::Type
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:60]: Please add a comment like the one above.

Comment (chriseth) [reply ast/src/parsed/mod.rs:60]: And maybe sort them so that the more complicated ones go first.

File: pil-analyzer/src/pil_analyzer.rs [hunk 17]
```diff
@@ -542,6 +560,40 @@ impl PILAnalyzer {
...
+                .filter_map(|(_, def)| def.as_ref())
+                .flat_map(|def| def.children())
+                .chain(
+                    self.trait_impls
+                        .values()
+                        .flat_map(|impls| impls.iter())
+                        .flat_map(|impl_| impl_.children()),
+                )
+                .chain(self.proof_items.iter()),
+        )
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:577]: Maybe also the publics?

Comment (chriseth) [reply pil-analyzer/src/pil_analyzer.rs:577]: Ah no it's fine.

File: pil-analyzer/src/structural_checks.rs [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+                duplicate_fields
+                    .into_iter()
+                    .map(|f| sr.with_error(format!("Field '{f}' specified more than once"))),
+            );
+        }
+
+        visited_structs.insert(name);
+    }
+
+    let structs_decls = definitions.iter().filter(|(_, (_, def))| {
```
Comment (chriseth) [inline pil-analyzer/src/structural_checks.rs:61]: Could be a new function starting here?
If you do that, please also create a wrapper function in this file so that we have only one call into this file from pil_analyzer.

File: pil-analyzer/src/structural_checks.rs [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+    parsed::{StructExpression, TypeDeclaration},
+};
+use powdr_parser_util::{Error, SourceRef};
+
+/// Verifies that all struct instantiations match their corresponding declarations
+/// (existence of field names, completeness) and ensures that both are correct.
+pub fn check_structs_fields<'a>(
+    structs_exprs: impl Iterator<Item = (&'a SourceRef, &'a StructExpression<Reference>)>,
+    definitions: &HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+) -> Vec<Error> {
```
Comment (chriseth) [inline pil-analyzer/src/structural_checks.rs:15]:
```suggestion
) -> Result<(), Vec<Error>> {
```

File: pil-analyzer/src/structural_checks.rs [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+                .duplicates()
+                .collect();
+
+            errors.extend(duplicate_declaration_fields.into_iter().map(|f| {
+                symbol
+                    .source
+                    .with_error(format!("Field '{f}' is declared more than once"))
+            }));
+        } else {
+            unreachable!()
```
Comment (chriseth) [inline pil-analyzer/src/structural_checks.rs:88]: In that case I would use a `let ... else {unreachable!()};`

File: pil-analyzer/src/structural_checks.rs [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+use powdr_parser_util::{Error, SourceRef};
+
+/// Verifies that all struct instantiations match their corresponding declarations
+/// (existence of field names, completeness) and ensures that both are correct.
+pub fn check_structs_fields<'a>(
+    structs_exprs: impl Iterator<Item = (&'a SourceRef, &'a StructExpression<Reference>)>,
+    definitions: &HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+) -> Vec<Error> {
+    let mut errors = Vec::new();
+    let mut visited_structs = HashSet::new();
```
Comment (chriseth) [inline pil-analyzer/src/structural_checks.rs:17]: Does not seem to be needed.

File: pil-analyzer/src/pil_analyzer.rs [hunk 22]
```diff
@@ -257,6 +258,23 @@ impl PILAnalyzer {
         }
     }
 
+    pub fn struct_fields_check(&self) -> Result<(), Vec<Error>> {
+        let structs_exprs = self.all_children().filter_map(|expr| {
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:262]: I'm wondering if it would be better to do that inside `check_structs_fields`?

File: pil-analyzer/src/structural_checks.rs [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+
+    for (sr, st) in structs_exprs {
```
Comment (chriseth) [inline pil-analyzer/src/structural_checks.rs:19]:
```suggestion
    for (source, st) in structs_exprs {
```

Comment (chriseth) [reply pil-analyzer/src/structural_checks.rs:19]: I didn't remember what it was

File: pil-analyzer/src/pil_analyzer.rs (hunk 22 above)
Comment (gzanitti) [reply pil-analyzer/src/pil_analyzer.rs:258]: Not sure, but the idea of having to pass `PILAnalyzer` (to make it possible to do `self.all_children()`), doesn't really convince me.

File: ast/src/parsed/mod.rs [hunk 24]
```diff
@@ -42,23 +42,24 @@ pub enum SymbolCategory {
     TypeConstructor,
     /// A trait declaration, which can be used as a type.
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:43]: Is this implemented like that? I actually don't think we currently do that, and for simplicity, I would prefer to keep it like `let<T: MyTrait<T>> f: T -> int` for now. We can introduce a shorthand like `let f: MyTrait -> int` later like rust does it.

File: pil-analyzer/src/structural_checks.rs [hunk 25]
```diff
@@ -25,39 +24,61 @@ pub fn check_structs_fields<'a>(
...
+}
+
+fn check_struct_expression(
+    definitions: &HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+    name: &String,
+    fields: &[powdr_ast::parsed::NamedExpression<Box<powdr_ast::parsed::Expression<Reference>>>],
+    source: &SourceRef,
+) -> Vec<Error> {
+    let mut errors = Vec::new();
+    if let (
```
Comment (chriseth) [inline pil-analyzer/src/structural_checks.rs:46]: Shouldn't this panic if it's not a `FunctionValueDefinition::TypeDeclaration`?

Comment (gzanitti) [reply pil-analyzer/src/structural_checks.rs:46]: Since names comes from StructExpressions, None shouldn't be a possible case, hence the direct unwrap.

File: pil-analyzer/src/structural_checks.rs [hunk 26]
```diff
@@ -25,39 +24,61 @@ pub fn check_structs_fields<'a>(
             unreachable!()
```
[the remark this answers is not in the document]
Comment (gzanitti) [reply pil-analyzer/src/structural_checks.rs:24]: I think it is mainly due to the `Reference::Poly(...)`. Which I understand is the only possible case, that's why the  `unreachable!()` else

File: pil-analyzer/src/structural_checks.rs [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+    for expr in structs_exprs {
+        let Expression::StructExpression(
+            source,
+            StructExpression {
+                name: Reference::Poly(PolynomialReference { name, .. }),
+                fields,
+            },
+        ) = expr
+        else {
+            unreachable!()
```
Comment (chriseth) [inline pil-analyzer/src/structural_checks.rs:27]: To be honest, instead of this I would prefer just filtering the expressions here instead of in the analyzer, but you can do this in a separate pr.

File: pil-analyzer/src/structural_checks.rs (hunk 25 above)
Comment (chriseth) [reply pil-analyzer/src/structural_checks.rs:49]: What I meant is that you have an `if let { ... }` here, but there is no else case. Shouldn't it panic in the else case?

File: pil-analyzer/src/structural_checks.rs [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+    errors
+}
+
+fn check_struct_declarations(
+    definitions: &HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+) -> Vec<Error> {
+    let structs_decls = definitions.iter().filter(|(_, (_, def))| {
+        matches!(
+            def,
+            Some(FunctionValueDefinition::TypeDeclaration(
```
Comment (chriseth) [inline pil-analyzer/src/structural_checks.rs:88]: Exactly the same pattern is matched 10 lines below. Why not iterate over all definitions, use `if let Some(FunctionValueDefinition::TypeDeclaration...) = def else { contine; }` ?

**Review by chriseth — APPROVED** [review]
